node_modules created in tomcat home/bin/

### Description of the bug

When I `mvn clean` the project and forget to run `mvn vaadin:prepare-frontend`, the `flow-build-info.json` file is missing. Now when I try to run the Skeleton starter in Tomcat from Intellij, the Vaadin Servlet autodetection magic misdetects the app home folder to `tomcat home/bin/` folder. Vaadin Servlet will thus starts populating `tomcat home/bin/` folder with `node_modules`, `webpack.config.js` etc, then it usually crashes that the `tomcat home/bin/frontend` folder doesn't contain referenced CSSs and the like.

The autodetection magic should instead give up in such cases and just fail with an informative exception.

In order to provide a better stacktrace I've disabled write access to my `tomcat home/bin/` folder. This also works as an excellent fail-fast fuse for these cases.

### Minimal reproducible example

1. git clone skeleton starter
2. run `mvn clean`
3. Run the Skeleton Starter in the Tomcat Local Server from Intellij
4. Observe the `tomcat home/bin` folder populated with `frontend`, `node_modules` etc.

### Expected behavior

An exception should have been thrown, instructing the user on how to add the missing `flow-build-info.json` file.

### Actual behavior

```
[2020-05-06 08:27:52,917] Artifact skeleton-starter-flow:war exploded: Artifact is being deployed, please wait...
06-May-2020 08:27:54.786 INFO [RMI TCP Connection(2)-127.0.0.1] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
06-May-2020 08:27:54.823 INFO [RMI TCP Connection(2)-127.0.0.1] org.apache.catalina.core.ApplicationContext.log Initializing AtmosphereFramework
[RMI TCP Connection(2)-127.0.0.1] INFO com.vaadin.flow.server.startup.DevModeInitializer - Starting dev-mode updaters in /home/mavi/local/apache-tomcat-9.0.31/bin folder.
06-May-2020 08:27:54.993 SEVERE [RMI TCP Connection(2)-127.0.0.1] org.apache.tomcat.util.modeler.BaseModelMBean.invoke Exception invoking method [manageApp]
	java.lang.IllegalStateException: Error starting child
		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:720)
		at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:690)
		at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:705)
		at org.apache.catalina.startup.HostConfig.manageApp(HostConfig.java:1728)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:289)
		at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
		at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
		at org.apache.catalina.mbeans.MBeanFactory.createStandardContext(MBeanFactory.java:456)
		at org.apache.catalina.mbeans.MBeanFactory.createStandardContext(MBeanFactory.java:405)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:289)
		at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
		at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
		at com.sun.jmx.remote.security.MBeanServerAccessController.invoke(MBeanServerAccessController.java:468)
		at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1468)
		at javax.management.remote.rmi.RMIConnectionImpl.access$300(RMIConnectionImpl.java:76)
		at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1309)
		at java.security.AccessController.doPrivileged(Native Method)
		at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1408)
		at javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:829)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:357)
		at sun.rmi.transport.Transport$1.run(Transport.java:200)
		at sun.rmi.transport.Transport$1.run(Transport.java:197)
		at java.security.AccessController.doPrivileged(Native Method)
		at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
		at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
		at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
		at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
		at java.security.AccessController.doPrivileged(Native Method)
		at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
		at java.lang.Thread.run(Thread.java:748)
	Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[]]
		at org.apache.catalina.util.LifecycleBase.handleSubClassException(LifecycleBase.java:440)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:198)
		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:717)
		... 43 more
	Caused by: java.io.UncheckedIOException: Failed to create directory '/home/mavi/local/apache-tomcat-9.0.31/bin/target/frontend'
		at com.vaadin.flow.server.startup.DevModeInitializer.initDevModeHandler(DevModeInitializer.java:235)
		at com.vaadin.flow.server.startup.DevModeInitializer.process(DevModeInitializer.java:182)
		at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.onStartup(ClassLoaderAwareServletContainerInitializer.java:76)
		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5135)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		... 44 more
	Caused by: java.io.IOException: Unable to create directory /home/mavi/local/apache-tomcat-9.0.31/bin/target/frontend
		at org.apache.commons.io.FileUtils.forceMkdir(FileUtils.java:2472)
		at com.vaadin.flow.server.startup.DevModeInitializer.initDevModeHandler(DevModeInitializer.java:232)
		... 48 more
06-May-2020 08:27:54.995 SEVERE [RMI TCP Connection(2)-127.0.0.1] org.apache.tomcat.util.modeler.BaseModelMBean.invoke Exception invoking method [createStandardContext]
	javax.management.RuntimeOperationsException: Exception invoking method [manageApp]
		at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:298)
		at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
		at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
		at org.apache.catalina.mbeans.MBeanFactory.createStandardContext(MBeanFactory.java:456)
		at org.apache.catalina.mbeans.MBeanFactory.createStandardContext(MBeanFactory.java:405)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:289)
		at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
		at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
		at com.sun.jmx.remote.security.MBeanServerAccessController.invoke(MBeanServerAccessController.java:468)
		at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1468)
		at javax.management.remote.rmi.RMIConnectionImpl.access$300(RMIConnectionImpl.java:76)
		at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1309)
		at java.security.AccessController.doPrivileged(Native Method)
		at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1408)
		at javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:829)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:357)
		at sun.rmi.transport.Transport$1.run(Transport.java:200)
		at sun.rmi.transport.Transport$1.run(Transport.java:197)
		at java.security.AccessController.doPrivileged(Native Method)
		at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
		at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
		at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
		at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
		at java.security.AccessController.doPrivileged(Native Method)
		at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
		at java.lang.Thread.run(Thread.java:748)
	Caused by: java.lang.IllegalStateException: Error starting child
		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:720)
		at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:690)
		at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:705)
		at org.apache.catalina.startup.HostConfig.manageApp(HostConfig.java:1728)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:289)
		... 35 more
	Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[]]
		at org.apache.catalina.util.LifecycleBase.handleSubClassException(LifecycleBase.java:440)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:198)
		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:717)
		... 43 more
	Caused by: java.io.UncheckedIOException: Failed to create directory '/home/mavi/local/apache-tomcat-9.0.31/bin/target/frontend'
		at com.vaadin.flow.server.startup.DevModeInitializer.initDevModeHandler(DevModeInitializer.java:235)
		at com.vaadin.flow.server.startup.DevModeInitializer.process(DevModeInitializer.java:182)
		at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.onStartup(ClassLoaderAwareServletContainerInitializer.java:76)
		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5135)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		... 44 more
	Caused by: java.io.IOException: Unable to create directory /home/mavi/local/apache-tomcat-9.0.31/bin/target/frontend
		at org.apache.commons.io.FileUtils.forceMkdir(FileUtils.java:2472)
		at com.vaadin.flow.server.startup.DevModeInitializer.initDevModeHandler(DevModeInitializer.java:232)
		... 48 more
[2020-05-06 08:27:55,024] Artifact skeleton-starter-flow:war exploded: Error during artifact deployment. See server log for details.
```

### Versions:
    - Vaadin / Flow version: 14.1.27
    - Java version: 8
    - OS version: Ubuntu x86-64 20.04
    - Browser version (if applicable): Firefox 75
    - Application Server (if applicable): Tomcat 9.0.31
    - IDE (if applicable): Intellij

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
@@ -267,8 +267,10 @@ public class DevModeInitializer
             return;
         }
 
-        String baseDir = config.getStringProperty(FrontendUtils.PROJECT_BASEDIR,
-                System.getProperty("user.dir", "."));
+        String baseDir = config.getStringProperty(FrontendUtils.PROJECT_BASEDIR, null);
+        if (baseDir == null) {
+            baseDir = getBaseDirectoryFallback();
+        }
 
         String generatedDir = System.getProperty(PARAM_GENERATED_DIR,
                 DEFAULT_GENERATED_DIR);
@@ -423,6 +425,28 @@ public class DevModeInitializer
         }
     }
 
+    /*
+     * Accept user.dir or cwd as a fallback only if the directory seems to be a
+     * Maven or Gradle project. Check to avoid cluttering server directories
+     * (see tickets #8249, #8403).
+     */
+    private static String getBaseDirectoryFallback() {
+        String baseDirCandidate = System.getProperty("user.dir", ".");
+        Path path = Paths.get(baseDirCandidate);
+        if (Files.isDirectory(path) && (Files.exists(path.resolve("pom.xml"))
+                || Files.exists(path.resolve("build.gradle")))) {
+            return path.toString();
+        } else {
+            throw new IllegalStateException(String.format(
+                    "Failed to determine project directory for dev mode. "
+                            + "Directory '%s' does not look like a Maven or "
+                            + "Gradle project. Ensure that you have run the "
+                            + "prepare-frontend Maven goal prior to deploying "
+                            + "your application",
+                    path.toString()));
+        }
+    }
+    
     /*
      * This method returns all folders of jar files having files in the
      * META-INF/resources/frontend folder. We don't use URLClassLoader because

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
@@ -442,8 +442,9 @@ public class DevModeInitializer
                     "Failed to determine project directory for dev mode. "
                             + "Directory '%s' does not look like a Maven or "
                             + "Gradle project. Ensure that you have run the "
-                            + "prepare-frontend Maven goal prior to deploying "
-                            + "your application",
+                            + "prepare-frontend Maven goal, which generates "
+                            +"'flow-build-info.json', prior to deploying your "
+                            + "application",
                     path.toString()));
         }
     }

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
@@ -433,8 +433,9 @@ public class DevModeInitializer
     private static String getBaseDirectoryFallback() {
         String baseDirCandidate = System.getProperty("user.dir", ".");
         Path path = Paths.get(baseDirCandidate);
-        if (Files.isDirectory(path) && (Files.exists(path.resolve("pom.xml"))
-                || Files.exists(path.resolve("build.gradle")))) {
+        if (path.toFile().isDirectory()
+                && (path.resolve("pom.xml").toFile().exists()
+                        || path.resolve("build.gradle").toFile().exists())) {
             return path.toString();
         } else {
             throw new IllegalStateException(String.format(

--- a/flow-server/src/test/java/com/vaadin/flow/server/startup/DevModeInitializerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/startup/DevModeInitializerTest.java
@@ -438,6 +438,27 @@ public class DevModeInitializerTest extends DevModeInitializerTestBase {
         }
     }
 
+    @Test
+    public void onStartup_fallbackBaseDirIsGradleProjectDirectory_isAccepted()
+            throws Exception {
+        initParams.remove(FrontendUtils.PROJECT_BASEDIR);
+        TemporaryFolder tmp = new TemporaryFolder();
+        tmp.create();
+        tmp.newFile("build.gradle");
+        baseDir = tmp.getRoot().getPath();
+
+        String originalUserDirValue = null;
+        try {
+            originalUserDirValue = System.getProperty("user.dir");
+            System.setProperty("user.dir", baseDir);
+            devModeInitializer.onStartup(classes, servletContext);
+        } finally {
+            if (originalUserDirValue != null) {
+                System.setProperty("user.dir", originalUserDirValue);
+            }
+        }
+    }
+
     private void loadingJars_allFilesExist(String resourcesFolder)
             throws IOException, ServletException {
         loadingJarsWithProtocol_allFilesExist(resourcesFolder, s -> {

--- a/flow-server/src/test/java/com/vaadin/flow/server/startup/DevModeInitializerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/startup/DevModeInitializerTest.java
@@ -29,6 +29,7 @@ import org.junit.Assert;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
 import org.mockito.ArgumentCaptor;
 import org.mockito.Mockito;
 
@@ -394,6 +395,47 @@ public class DevModeInitializerTest extends DevModeInitializerTestBase {
                         servletContextAttributes.get(answer.getArgumentAt(0, String.class)));
         process();
         assertTrue(DevModeInitializer.isDevModeAlreadyStarted(servletContext));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void onStartup_fallbackBaseDirIsNotProjectDirectory_throws()
+            throws Exception {
+        initParams.remove(FrontendUtils.PROJECT_BASEDIR);
+        TemporaryFolder tmp = new TemporaryFolder();
+        tmp.create();
+        baseDir = tmp.getRoot().getPath();
+
+        String originalUserDirValue = null;
+        try {
+            originalUserDirValue = System.getProperty("user.dir");
+            System.setProperty("user.dir", baseDir);
+            devModeInitializer.onStartup(classes, servletContext);
+        } finally {
+            if (originalUserDirValue != null) {
+                System.setProperty("user.dir", originalUserDirValue);
+            }
+        }
+    }
+
+    @Test
+    public void onStartup_fallbackBaseDirIsMavenProjectDirectory_isAccepted()
+            throws Exception {
+        initParams.remove(FrontendUtils.PROJECT_BASEDIR);
+        TemporaryFolder tmp = new TemporaryFolder();
+        tmp.create();
+        tmp.newFile("pom.xml");
+        baseDir = tmp.getRoot().getPath();
+
+        String originalUserDirValue = null;
+        try {
+            originalUserDirValue = System.getProperty("user.dir");
+            System.setProperty("user.dir", baseDir);
+            devModeInitializer.onStartup(classes, servletContext);
+        } finally {
+            if (originalUserDirValue != null) {
+                System.setProperty("user.dir", originalUserDirValue);
+            }
+        }
     }
 
     private void loadingJars_allFilesExist(String resourcesFolder)

